util/sock_util.c

```diff
@@ -276,3 +278,67 @@ sock_addr_prettyprint(const struct sock_addr * sa)
 		return (strdup("Unknown address"));
 	}
 }
+
+/**
+ * sock_addr_ensure_port(addr):
+ * Allocate a new string to serve as the address for sock_resolve().
+ * If ${addr} contains a port number or is the address of a Unix domain
+ * socket, duplicate that string; if not, add a port number of ":0".
+ */
+char *
+sock_addr_ensure_port(const char * addr)
+{
+	char * bind_addr;
+	char * cl;
+	char * cr;
+
+	/* Sanity check. */
+	assert(addr != NULL);
+
+	/* Where are the left-most and right-most colons in $addr? */
+	cl = strchr(addr, ':');
+	cr = strrchr(addr, ':');
+
+	/* If the right-most colon is the left-most character, bail. */
+	if (cr == addr) {
+		warn0("Invalid address: %s", addr);
```

Comment:
I'm inclined to strdup the address here.  The string we return won't be a valid address, but it won't be any more invalid than the one we were given; let the caller decide how to report invalid addresses.